parsl/providers/ad_hoc/ad_hoc.py

```diff
@@ -0,0 +1,234 @@
+import logging
+import os
+import time
+
+from parsl.channels import LocalChannel
+from parsl.launchers import SimpleLauncher
+from parsl.providers.provider_base import ExecutionProvider
+from parsl.providers.error import SchedulerMissingArgs, ScriptPathError
+from parsl.utils import RepresentationMixin
+
+logger = logging.getLogger(__name__)
+
+
+translate_table = {
+    'R': 'RUNNING',
+    'CA': 'CANCELLED',
+    'F': 'FAILED',  # (failed),
+}  # (special exit state
+
+
+def _roundrobin(items):
+    """ Returns items one at a time in an infinite loop
+    """
+    while True:
+        try:
+            for item in items:
+                yield item
+        except Exception:
+            # Remove the iterator we just exhausted from the cycle.
+            pass
+
+
+class AdHocProvider(ExecutionProvider, RepresentationMixin):
+    """ Ad-hoc execution provider
+
+    This provider is used to provision execution resources over one or more ad hoc nodes
+    that are accessible over a Channel (say, ssh) but otherwise lack a cluster scheduler.
+
+    Each submit call invoked will go through the list channels in a round-robin fashion
+
+    Parameters
+    ----------
+
+    worker_init : str
+      Command to be run before starting a worker, such as 'module load Anaconda; source activate env'.
+      Since this provider calls the same worker_init across all nodes in the ad-hoc cluster, it is
+      recommended that a single script is made available across nodes such as ~/setup.sh that can
+      be invoked.
+
+    """
+
+    def __init__(self,
+                 channels=[],
+                 worker_init='',
+                 cmd_timeout=30,
+                 parallelism=1,
+                 init_blocks=0,
```

Comment:
`init_blocks` should default to 1?

Reply:
In this case it should be set by the user to # of channels.

Reply:
Should *always* ?

If so, then maybe the three `*_blocks` parameters shouldn't be parameters at all and they should all be set to the length of the channels attribute.

Reply:
A bit confused here. Maybe they should not be available to configure?

And it seems there should not be any scaling for AdHocProvider

Reply:
Or it should scale in terms of channels/nodes.

Reply:
I'll have this be set internally to `init_blocks = min_blocks = max_blocks = #of channels`